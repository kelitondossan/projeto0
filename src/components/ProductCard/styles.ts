import styled, {css, keyframes} from "styled-components";

export const CardContainer = styled.div`
    width: 100%;
    display: grid;
    grid-template-columns: repeat(4, 1fr);
    gap: 1.25rem;

    @media (max-width: 1080px) {
        grid-template-columns: repeat(2, 1fr);
        gap: 1.5rem;
    }
    @media (max-width: 768px) {
        grid-template-columns: repeat(1, 1fr);
        gap: 1.5rem;
    }
`;

export const Card = styled.div`
    display: flex;
    align-items: center;
    flex-direction: column;
    justify-content: center;
    position: relative;
    padding: 1.125rem 0.5rem 3rem;
    background: #FFF;
    border-radius: 8px;
    box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.135216);
    span{
        font-size: 15px;
        line-height: 19px;
    }
    span:last-child{
        padding: 0.25rem 0.5rem;
        background-color: #373737;
        color: white;
        font-weight: 700;
        border-radius: 5px;
        
    }
    img{
        object-fit: contain;
    }

    p{
        margin-top: 0.5rem;
        font-size: 10px;
        font-weight: 300;
        line-height: 12px;
        color: #2C2C2C;
        
    }
`;

export const CarDesc = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-between;
    width: 100%;
`;

export const CardButton = styled.button`
    padding: 0.5rem 0;
    position: absolute;
    bottom: 0;
    
    display: flex;
    justify-content: center;

    font-size: 14px;
    width: 100%;
    border: none;
    outline: none;
    background: #0F52BA;
    color: #fff;
    text-transform: uppercase;
    font-weight: 600;
    cursor: pointer;
    border-radius: 0px 0px 8px 8px;

    svg{
        margin-right: 8px;

    }

`;  

const shimmer = css`
    background: #f6f7f8
    linear-gradient(
      to right,
      #f6f7f8 0%,
      #edeef1 20%,
      #f6f7f8 40%,
      #ecebeb 100%
    )
    no-repeat;
  background-size: 800px 800px;
  display: inline-block;
  position: relative;
  border-radius: 5px;

  animation: shimmer 1.2s ease-in-out infinite;

  @keyframes shimmer {
    0% {
      background-position: 0% 0%;
    }
    100%{
      background-position: -135% 0%;
    }
  }
`;

export const Skeleton = styled.div`
    display: flex;
    //border: 1px solid red;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    background: var(--white);
    gap: 10px;
    padding: 1.125rem 0.5rem 1.25rem;
    border-radius: 5px;
    position: relative;
`;

export const Box = styled.div`
  ${shimmer};
  display: flex;
  width: 150px;
  height: 150px;
  margin-bottom: 10px;
  border-radius: 5px;
`;

export const Text = styled.div`
  ${shimmer};
  width: 100%;
  height: 15px;
  border-radius: 0 0 5px 5px;
  :last-child{
    position: absolute;
    bottom: 0;
  }
  
`;