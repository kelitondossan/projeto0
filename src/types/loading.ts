interface LoadingProps {
    isloading: boolean
}

export default LoadingProps;