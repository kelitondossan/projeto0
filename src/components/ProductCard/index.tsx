import React, { useEffect, useState } from "react";
import api from "../../services/api";
import {
  CardContainer,
  Card,
  CarDesc,
  CardButton,
  Box,
  Text,
  Skeleton,
} from "./styles";
import { Product } from "@/types/products";
import { FiShoppingBag } from "react-icons/fi";
import { addToCart } from "../../features/cartSlice";
import { useDispatch } from "react-redux";

interface ProductData {
  products: Product[];
}

const ProductCard: React.FC = () => {
  const dispatch = useDispatch();
  const [isLoading, setIsLoading] = useState(true);
  const [products, setProducts] = useState<ProductData["products"]>([]);

  useEffect(() => {
    setTimeout(() => {
      setIsLoading(false);
    }, 1500);
    const fetchData = async () => {
      const result = await api.get(
        `products?page=1&rows=8&sortBy=name&orderBy=DESC`
      );
      setProducts(result.data.products);
    };
    fetchData();
  }, []);

  return (
    <>
      <CardContainer>
        {!isLoading
          ? products.map((product, index) => (
              <Card key={product.id}>
                <img src={product.photo} width={150} height={150}></img>
                <CarDesc>
                  <span>{product.name}</span>
                  <span>
                    R$
                    {parseFloat(product.price.toString().replace(",", ""))}
                  </span>
                </CarDesc>
                <p>{product.description}</p>
                <CardButton onClick={() => dispatch(addToCart(product))}>
                  <FiShoppingBag />
                  Comprar
                </CardButton>
              </Card>
            ))
          : products.map((product) => <CardSkeleton key={product.id} />)}
      </CardContainer>
    </>
  );
};

export const CardSkeleton = () => {
  return (
    <Skeleton>
      <Box />
      <Text />
      <Text />
      <Text />
    </Skeleton>
  );
};

export default ProductCard;
