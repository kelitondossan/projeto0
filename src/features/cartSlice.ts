import { Product } from "@/types/products";
import { bindActionCreators, createSlice } from "@reduxjs/toolkit";
import { store } from "./store";
import { ShowToast } from "../components/ToastMessage";

type ProductState = {
    cartItems: Array<Product>,
    totalQuantity: Number,
    totalPrice: number
};

const initialState = {
    cartItems: [],
    totalPrice: 0,
    totalQuantity: 0
} as ProductState;

const CartSlice = createSlice({
    name: "cart",
    initialState,
    reducers:{
        addToCart: (state, action) => {
            const addItem = state.cartItems.find(
                (Cartitem) => Cartitem.id === action.payload.id
            );
            addItem ? (addItem.quantity = addItem.quantity + 1 ) : state.cartItems.push({...action.payload, quantity: 1});
            
            ShowToast({type: 'success', message: `${action.payload.name} No carrinho`})
        },
        increase: (state, action) => {
            
            const itemIndex = state.cartItems.findIndex(
                (cartItem) => cartItem.id === action.payload.id
            );
            //if (itemIndex === -1) return;
            state.cartItems[itemIndex].quantity++;
        },
        decrease: (state, action) =>{
            const itemIndex = state.cartItems.findIndex(
                (cartItem) => cartItem.id === action.payload.id
            );
            state.cartItems[itemIndex].quantity--;
            if (state.cartItems[itemIndex].quantity === 0) {
                state.cartItems.splice(itemIndex, 1)
            } else {
                state.cartItems[itemIndex].quantity < 1;
            };
        },
        removeToCart: (state, action) =>{
            state.cartItems =  state.cartItems.filter((cartItem) => cartItem.id !== action.payload);
        },

        getCartTotal: (state, action) => {
            let { total, quantity } = state.cartItems.reduce(
                (cartTotal, cartItem) =>{
                    const { price, quantity } = cartItem;
                    const itemTotal = price * quantity;

                    cartTotal.total += itemTotal;
                    cartTotal.quantity += quantity;

                    return cartTotal;
                },
                {
                    total: 0,
                    quantity: 0
                }
            );
            state.totalQuantity = quantity;
            state.totalPrice = total;
        },
        
        clear: (state, action) => {
            state.cartItems = state.cartItems.filter((cartItem) => cartItem.id !== action.payload);
        }
    }
})
export type initialState = ReturnType<typeof store.getState>
export const { addToCart, increase, decrease, removeToCart, clear, getCartTotal } = CartSlice.actions;
export default CartSlice.reducer