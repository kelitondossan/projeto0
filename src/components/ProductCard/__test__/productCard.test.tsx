import { Product } from "@/types/products";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import ProductCard from "..";
import { store } from "../../../features/store";

describe("productCard Component", () => {
  it("should render correctly", () => {
    render(
      <Provider store={store}>
        <ProductCard />
      </Provider>
    );
    expect(ProductCard).toMatchSnapshot();
  });
});
