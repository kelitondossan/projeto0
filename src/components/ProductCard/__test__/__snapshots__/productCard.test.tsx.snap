// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`productCard Component should render correctly 1`] = `[Function]`;
